File: app/javascript/channels/room_channel.js
import consumer from "./consumer"

// $(function() {}; で囲むことでレンダリング後に実行される
// レンダリング前に実行されると $('#messages').data('room_id') が取得できない
$(function() {
  const chatChannel = consumer.subscriptions.create({ channel: 'RoomChannel', room: $('#messages').data('room_id') }, {
    connected() {
      // Called when the subscription is ready for use on the server
    },

    disconnected() {
      // Called when the subscription has been terminated by the server
    },

    received: function(data) {
      // Called when there's incoming data on the websocket for this channel
      this.perform('received', {data});
      return $('#messages').append(data['message']);
    },

    // これが実行されるとコンシューマになったRoomChannel#speak({ message: message })が呼ばれる
    speak: function(message) {
      const current_user = $('.current_user').val();
      const current_user_id = $('.current_user_id').val();
      const current_user_name = $('.current_user_name').val();
      return this.perform('speak', {
        message: message,
        current_user: current_user,
        current_user_id: current_user_id,
        current_user_name: current_user_name
      });
    }
  });

  $(document).on('keypress', '[data-behavior~=room_speaker]', function(event) {
    if (event.keyCode === 13) {
      chatChannel.speak(event.target.value);
      event.target.value = '';
      return event.preventDefault();
    }
  });
});

File: app/channels/room_channel.rb
class RoomChannel < ApplicationCable::Channel
  def subscribed
    #オープンチャット機能用
    #stream_from "room_channel"
    #DM機能用
    stream_from "room_channel_#{params['room']}" 
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def receive(data)
    #ActionCable.server.broadcast("room_channel_#{params['room']}", data)
  end

  def speak(data)
    #オープンチャット機能用
    #Message.create! content: data['message'], user_id: data['current_user_id'], user_name: data['current_user_name']
    #DM機能用
    #Message.create! content: data['message'], user_id: data['current_user_id'], user_name: data['current_user_name'], room_id: params['room']
    #@user = data['current_user']
    #puts @user
    #puts current_user
    puts "oooooooooooooo"
    #puts current_user.id
    
    messages = current_user.messages.create(content: data['message'], user_id: data['current_user_id'], user_name: data['current_user_name'], room_id: params['room'])
    ActionCable.server.broadcast("room_channel_#{params['room']}", message: render_message(messages))
  end

  private

    def render_message(message)
      puts "AAAAAAAAAAAAAAAAAA"
      ApplicationController.renderer.render partial: 'messages/message', locals: { message: message }
    end
end
